# Maandalizi ya Data

[Chanzo cha Notebook asili kutoka *Data Science: Utangulizi wa Kujifunza Mashine kwa Data Science Python na Machine Learning Studio na Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## Kuchunguza taarifa za `DataFrame`

> **Lengo la kujifunza:** Mwisho wa sehemu hii, unapaswa kuwa na uelewa wa jinsi ya kupata taarifa za jumla kuhusu data iliyohifadhiwa katika pandas DataFrames.

Baada ya kupakia data yako kwenye pandas, kuna uwezekano mkubwa kuwa itakuwa katika `DataFrame`. Hata hivyo, ikiwa seti ya data katika `DataFrame` yako ina safu 60,000 na nguzo 400, utaanzaje kupata picha ya kile unachofanya kazi nacho? Kwa bahati nzuri, pandas inatoa zana rahisi za kuangalia haraka taarifa za jumla kuhusu `DataFrame` pamoja na safu chache za mwanzo na za mwisho.

Ili kuchunguza utendaji huu, tutapakia maktaba ya Python scikit-learn na kutumia seti ya data maarufu ambayo kila mtaalamu wa data ameiona mara nyingi: seti ya data ya *Iris* ya mwanabiolojia wa Uingereza Ronald Fisher iliyotumika katika karatasi yake ya mwaka 1936 "Matumizi ya vipimo vingi katika matatizo ya taxonomia":


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
Tumeweka Dataset ya Iris kwenye kigezo `iris_df`. Kabla ya kuanza kuchambua data, itakuwa muhimu kujua idadi ya vipengele tulivyonavyo na ukubwa wa jumla wa dataset. Ni muhimu kuangalia kiasi cha data tunachoshughulikia.


In [2]:
iris_df.shape

(150, 4)

Kwa hivyo, tunashughulika na safu 150 na nguzo 4 za data. Kila safu inawakilisha nukta moja ya data na kila nguzo inawakilisha kipengele kimoja kinachohusiana na fremu ya data. Kwa msingi, kuna nukta 150 za data zenye vipengele 4 kila moja.

`shape` hapa ni sifa ya fremu ya data na si kazi, ndiyo sababu haimalizii na jozi ya mabano.


### `DataFrame.columns`
Sasa tuingie kwenye safu 4 za data. Kila moja inawakilisha nini hasa? Sifa ya `columns` itatupa majina ya safu katika dataframe.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

Kama tunavyoona, kuna safu nne (4). Sifa ya `columns` inatuambia majina ya safu na kimsingi hakuna kingine. Sifa hii inakuwa muhimu tunapotaka kutambua vipengele ambavyo seti ya data inavyo.


### `DataFrame.info`
Kiasi cha data (kinachotolewa na sifa ya `shape`) na majina ya vipengele au safu (vinavyotolewa na sifa ya `columns`) vinatupa taarifa fulani kuhusu seti ya data. Sasa, tungependa kuchunguza kwa undani zaidi seti ya data. Kazi ya `DataFrame.info()` ni muhimu sana kwa hili.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


Kutoka hapa, tunaweza kufanya uchunguzi kadhaa:  
1. Aina ya Data ya kila safu: Katika seti hii ya data, data yote imehifadhiwa kama nambari za desimali za biti 64.  
2. Idadi ya Thamani Zisizo Null: Kushughulikia thamani za null ni hatua muhimu katika maandalizi ya data. Hili litatatuliwa baadaye kwenye daftari.  


### DataFrame.describe()
Tuseme tuna data nyingi za namba kwenye seti yetu ya data. Mahesabu ya takwimu za upande mmoja kama vile wastani, mediani, robo n.k. yanaweza kufanywa kwenye kila safu moja moja. Kazi ya `DataFrame.describe()` inatupatia muhtasari wa takwimu za safu za namba kwenye seti ya data.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

Matokeo hapo juu yanaonyesha jumla ya idadi ya pointi za data, wastani, upotofu wa kawaida, thamani ya chini, robo ya chini (25%), mediani (50%), robo ya juu (75%) na thamani ya juu ya kila safu.


### `DataFrame.head`
Kwa kutumia kazi na sifa zote zilizotajwa hapo juu, tumepata muhtasari wa juu wa seti ya data. Tunajua ni alama ngapi za data zipo, ni sifa ngapi zipo, aina ya data ya kila sifa, na idadi ya thamani zisizo tupu kwa kila sifa.

Sasa ni wakati wa kuangalia data yenyewe. Hebu tuone jinsi safu chache za mwanzo (alama chache za mwanzo za data) za `DataFrame` yetu zinavyoonekana:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

Kama matokeo hapa, tunaweza kuona maingizo matano (5) ya seti ya data. Tukitazama faharasa upande wa kushoto, tunagundua kuwa hizi ni safu tano za kwanza.


### Zoezi:

Kutoka kwa mfano uliotolewa hapo juu, ni wazi kwamba, kwa chaguo-msingi, `DataFrame.head` inarudisha safu tano za kwanza za `DataFrame`. Katika seli ya msimbo hapa chini, unaweza kugundua njia ya kuonyesha zaidi ya safu tano?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
Njia nyingine ya kuangalia data inaweza kuwa kutoka mwisho (badala ya mwanzo). Kinyume cha `DataFrame.head` ni `DataFrame.tail`, ambayo inarudisha safu tano za mwisho za `DataFrame`:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

Katika mazoezi, ni muhimu kuwa na uwezo wa kuchunguza kwa urahisi safu chache za mwanzo au safu chache za mwisho za `DataFrame`, hasa unapokuwa unatafuta thamani zisizo za kawaida katika seti za data zilizo na mpangilio.

Kazi zote na sifa zilizoonyeshwa hapo juu kwa msaada wa mifano ya msimbo, zinatusaidia kupata muonekano na hisia ya data.

> **Mafunzo:** Hata kwa kuangalia tu metadata kuhusu taarifa katika `DataFrame` au thamani chache za mwanzo na mwisho ndani yake, unaweza kupata wazo la haraka kuhusu ukubwa, umbo, na maudhui ya data unayoshughulikia.


### Data Inayokosekana
Hebu tujadili kuhusu data inayokosekana. Data inayokosekana hutokea pale ambapo hakuna thamani iliyohifadhiwa katika baadhi ya safu.

Hebu tuchukue mfano: tuseme mtu fulani anajali sana kuhusu uzito wake na hawezi kujaza sehemu ya uzito katika dodoso. Basi, thamani ya uzito kwa mtu huyo itakuwa haipo.

Mara nyingi, katika seti za data za ulimwengu halisi, thamani zinazokosekana hutokea.

**Jinsi Pandas Inavyoshughulikia Data Inayokosekana**

Pandas hushughulikia thamani zinazokosekana kwa njia mbili. Ya kwanza umeiona hapo awali katika sehemu zilizopita: `NaN`, au Not a Number. Hii ni thamani maalum ambayo ni sehemu ya vipimo vya IEEE vya namba za desimali na hutumika tu kuonyesha thamani za desimali zinazokosekana.

Kwa thamani zinazokosekana ambazo si za desimali, pandas hutumia kitu cha Python `None`. Ingawa inaweza kuonekana kuchanganya kwamba utakutana na aina mbili tofauti za thamani zinazomaanisha kimsingi kitu kimoja, kuna sababu za kimipangilio za kubuni chaguo hili, na kwa vitendo, njia hii inaiwezesha pandas kutoa suluhisho bora kwa hali nyingi. Hata hivyo, `None` na `NaN` zote zina vizuizi ambavyo unapaswa kuwa makini navyo kuhusiana na jinsi zinavyoweza kutumika.


### `None`: data isiyokuwepo isiyo ya aina ya float
Kwa sababu `None` inatoka kwenye Python, haiwezi kutumika katika arrays za NumPy na pandas ambazo si za aina ya data `'object'`. Kumbuka, arrays za NumPy (na miundo ya data katika pandas) zinaweza kuwa na aina moja tu ya data. Hii ndiyo inazipa nguvu kubwa kwa kazi za data na mahesabu makubwa, lakini pia inapunguza uwezo wao wa kubadilika. Arrays kama hizo lazima zibadilishwe hadi kwenye “kiwango cha chini cha kawaida,” yaani aina ya data ambayo itajumuisha kila kitu kilichopo kwenye array. Wakati `None` ipo kwenye array, inamaanisha unafanya kazi na vitu vya Python.

Ili kuona hili likifanyika, zingatia mfano wa array ifuatayo (angalia `dtype` yake):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

Uhalisia wa aina za data zilizopandishwa huleta athari mbili. Kwanza, operesheni zitafanyika katika kiwango cha msimbo wa Python unaotafsiriwa badala ya msimbo wa NumPy uliokusanywa. Kimsingi, hii inamaanisha kuwa operesheni yoyote inayohusisha `Series` au `DataFrames` zenye `None` ndani yake itakuwa polepole zaidi. Ingawa huenda usione athari hii ya utendaji, kwa seti kubwa za data inaweza kuwa tatizo.

Athari ya pili inatokana na ya kwanza. Kwa sababu `None` kimsingi inarudisha `Series` au `DataFrame`s katika ulimwengu wa Python ya kawaida, kutumia mkusanyiko wa NumPy/pandas kama `sum()` au `min()` kwenye safu ambazo zina thamani ya ``None`` kwa ujumla kutasababisha hitilafu:


In [10]:
example1.sum()

TypeError: ignored

### `NaN`: thamani za float zinazokosekana

Tofauti na `None`, NumPy (na kwa hivyo pandas) inasaidia `NaN` kwa ajili ya operesheni zake za haraka, za vekta, na ufuncs. Habari mbaya ni kwamba hesabu yoyote inayofanywa kwenye `NaN` daima husababisha `NaN`. Kwa mfano:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

Habari njema: mkusanyiko unaoendeshwa kwenye safu zenye `NaN` ndani yake hauleti makosa. Habari mbaya: matokeo si ya manufaa kwa usawa:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

In [11]:
# What happens if you add np.nan and None together?


### `NaN` na `None`: thamani tupu katika pandas

Ingawa `NaN` na `None` zinaweza kuonyesha tabia tofauti kidogo, pandas imeundwa kushughulikia zote mbili kwa njia inayofanana. Ili kuelewa tunachomaanisha, fikiria `Series` ya nambari kamili:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


Katika mchakato wa kubadilisha aina za data ili kuanzisha usawa wa data katika `Series` na `DataFrame`s, pandas itabadilisha kwa hiari thamani zilizokosekana kati ya `None` na `NaN`. Kwa sababu ya kipengele hiki cha muundo, inaweza kuwa muhimu kufikiria `None` na `NaN` kama ladha mbili tofauti za "null" katika pandas. Kwa kweli, baadhi ya mbinu za msingi utakazotumia kushughulikia thamani zilizokosekana katika pandas zinaakisi wazo hili katika majina yao:

- `isnull()`: Hutengeneza mask ya Boolean inayoonyesha thamani zilizokosekana
- `notnull()`: Kinyume cha `isnull()`
- `dropna()`: Hurejesha toleo lililochujwa la data
- `fillna()`: Hurejesha nakala ya data na thamani zilizokosekana zikiwa zimejazwa au kuwekewa makadirio

Hizi ni mbinu muhimu za kuzifahamu na kuzoea, kwa hivyo hebu tuzipitie kila moja kwa undani zaidi.


### Kugundua thamani za null

Sasa kwa kuwa tumeelewa umuhimu wa thamani zinazokosekana, tunahitaji kuzitambua kwenye seti yetu ya data kabla ya kushughulika nazo.  
Zote `isnull()` na `notnull()` ni mbinu zako za msingi za kugundua data ya null. Zote zinarudisha maski za Boolean juu ya data yako.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Angalia kwa makini matokeo. Je, kuna lolote linalokushangaza? Ingawa `0` ni null ya hesabu, bado ni nambari kamili nzuri na pandas inaitendea hivyo. `''` ni kidogo zaidi ya hila. Ingawa tulitumia katika Sehemu ya 1 kuwakilisha thamani ya mnyororo tupu, bado ni kitu cha mnyororo na si uwakilishi wa null kulingana na pandas.

Sasa, hebu tugeuze hili na tutumie mbinu hizi kwa namna ambayo utazitumia kwa vitendo. Unaweza kutumia vinyago vya Boolean moja kwa moja kama ``Series`` au ``DataFrame`` index, ambayo inaweza kuwa muhimu unapojaribu kufanya kazi na thamani zilizokosekana (au zilizopo) pekee.

Ikiwa tunataka jumla ya idadi ya thamani zilizokosekana, tunaweza kufanya jumla juu ya kinyago kinachozalishwa na mbinu ya `isnull()`.


In [19]:
example3.isnull().sum()

2

In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**Mambo ya msingi**: Mbinu za `isnull()` na `notnull()` hutoa matokeo yanayofanana unapotumia katika DataFrames: zinaonyesha matokeo na faharasa ya matokeo hayo, ambayo yatakusaidia sana unaposhughulika na data yako.


### Kushughulikia data iliyokosekana

> **Lengo la kujifunza:** Kufikia mwisho wa sehemu hii ndogo, unapaswa kujua jinsi na wakati wa kubadilisha au kuondoa thamani za null kutoka kwa DataFrames.

Mifano ya Kujifunza Mashine haziwezi kushughulikia data iliyokosekana zenyewe. Kwa hivyo, kabla ya kupitisha data kwenye mfano, tunahitaji kushughulikia thamani hizi zilizokosekana.

Jinsi data iliyokosekana inavyoshughulikiwa huleta maamuzi yenye athari ndogo, inaweza kuathiri uchambuzi wako wa mwisho na matokeo halisi ya ulimwengu.

Kuna njia mbili kuu za kushughulikia data iliyokosekana:

1.   Kuondoa safu inayojumuisha thamani iliyokosekana
2.   Kubadilisha thamani iliyokosekana na thamani nyingine

Tutajadili njia hizi zote mbili na faida na hasara zake kwa undani.


### Kuondoa Thamani Zisizo na Maana (Null Values)

Kiasi cha data tunachopitisha kwenye modeli yetu kina athari ya moja kwa moja kwenye utendaji wake. Kuondoa thamani zisizo na maana kunamaanisha tunapunguza idadi ya vipengele vya data, na hivyo kupunguza ukubwa wa seti ya data. Kwa hivyo, inashauriwa kuondoa safu zilizo na thamani zisizo na maana wakati seti ya data ni kubwa sana.

Mfano mwingine unaweza kuwa kwamba safu fulani au safu wima ina idadi kubwa ya thamani zinazokosekana. Katika hali hiyo, zinaweza kuondolewa kwa sababu hazitachangia sana kwenye uchambuzi wetu kwani data nyingi inakosekana kwa safu/safu wima hiyo.

Zaidi ya kutambua thamani zinazokosekana, pandas inatoa njia rahisi ya kuondoa thamani zisizo na maana kutoka kwa `Series` na `DataFrame`s. Ili kuona hili likifanyika, hebu turudi kwenye `example3`. Kazi ya `DataFrame.dropna()` husaidia kuondoa safu zilizo na thamani zisizo na maana.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Kumbuka kwamba hii inapaswa kuonekana kama matokeo yako kutoka `example3[example3.notnull()]`. Tofauti hapa ni kwamba, badala ya kuorodhesha tu kwenye thamani zilizofichwa, `dropna` imeondoa zile thamani zilizokosekana kutoka kwa `Series` `example3`.

Kwa sababu DataFrames zina vipimo viwili, zinatoa chaguo zaidi za kuondoa data.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(Je, uliona kwamba pandas ilibadilisha safu mbili kuwa nambari za desimali ili kushughulikia `NaN`?)

Huwezi kuondoa thamani moja pekee kutoka kwa `DataFrame`, kwa hivyo unapaswa kuondoa safu nzima au safu wima. Kulingana na unachofanya, unaweza kutaka kufanya moja au nyingine, na hivyo pandas inakupa chaguo kwa zote mbili. Kwa sababu katika sayansi ya data, safu wima kwa kawaida huwakilisha vigezo na safu huwakilisha uchunguzi, kuna uwezekano mkubwa wa kuondoa safu za data; mpangilio wa chaguo-msingi wa `dropna()` ni kuondoa safu zote ambazo zina thamani yoyote isiyo na maana:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

Ikiwa ni lazima, unaweza kuondoa thamani za NA kutoka kwenye safu. Tumia `axis=1` kufanya hivyo:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

Kumbuka kwamba hii inaweza kuondoa data nyingi ambayo unaweza kutaka kuhifadhi, hasa katika seti ndogo za data. Je, unafanyaje ikiwa unataka tu kuondoa safu au nguzo ambazo zina thamani kadhaa au hata zote zikiwa tupu? Unaweza kubainisha mipangilio hiyo katika `dropna` kwa kutumia vigezo vya `how` na `thresh`.

Kwa chaguo-msingi, `how='any'` (ikiwa ungependa kuthibitisha mwenyewe au kuona vigezo vingine ambavyo njia hii inayo, endesha `example4.dropna?` katika seli ya msimbo). Vinginevyo, unaweza kubainisha `how='all'` ili kuondoa tu safu au nguzo ambazo zina thamani zote tupu. Hebu tuongeze mfano wetu wa `DataFrame` ili kuona hili likifanyika katika zoezi lijalo.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> Mambo ya Muhimu:  
1. Kuondoa thamani za null ni wazo zuri tu ikiwa seti ya data ni kubwa vya kutosha.  
2. Safu nzima au nguzo zinaweza kuondolewa ikiwa zina data nyingi zilizokosekana.  
3. Njia ya `DataFrame.dropna(axis=)` husaidia kuondoa thamani za null. Hoja ya `axis` inaonyesha kama safu (rows) zinapaswa kuondolewa au nguzo (columns).  
4. Hoja ya `how` pia inaweza kutumika. Kwa chaguo-msingi imewekwa kuwa `any`. Hivyo, inaondoa tu safu/nguzo ambazo zina thamani yoyote ya null. Inaweza kuwekwa kuwa `all` ili kubainisha kwamba tutaondoa tu safu/nguzo ambazo thamani zote ni null.  


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


Kigezo cha `thresh` kinakupa udhibiti wa kina zaidi: unaweka idadi ya thamani *zisizo tupu* ambazo safu au safu wima inahitaji kuwa nazo ili kuhifadhiwa:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

### Kujaza Thamani Zilizokosekana

Wakati mwingine inafaa kujaza thamani zilizokosekana na zile ambazo zinaweza kuwa halali. Kuna mbinu kadhaa za kujaza thamani tupu. Ya kwanza ni kutumia Maarifa ya Eneo (maarifa ya mada ambayo dataset inahusu) ili kwa namna fulani kukadiria thamani zilizokosekana.

Unaweza kutumia `isnull` kufanya hili moja kwa moja, lakini hilo linaweza kuwa kazi ngumu, hasa ikiwa una thamani nyingi za kujaza. Kwa sababu hii ni kazi ya kawaida sana katika sayansi ya data, pandas inatoa `fillna`, ambayo inarudisha nakala ya `Series` au `DataFrame` na thamani zilizokosekana kubadilishwa na zile unazochagua. Hebu tuunde mfano mwingine wa `Series` ili kuona jinsi hii inavyofanya kazi kwa vitendo.


### Data ya Kategoria (Isiyo ya Nambari)
Kwanza, tuzingatie data isiyo ya nambari. Katika seti za data, tuna safu zenye data ya kategoria. Mfano: Jinsia, Kweli au Siyo Kweli, n.k.

Katika hali nyingi, tunabadilisha thamani zilizokosekana kwa kutumia `mode` ya safu husika. Kwa mfano, tuna alama 100 za data ambapo 90 zimesema Kweli, 8 zimesema Siyo Kweli, na 2 hazijajazwa. Basi, tunaweza kujaza zile 2 kwa Kweli, tukizingatia safu nzima.

Tena, hapa tunaweza kutumia maarifa ya uwanja. Hebu tuzingatie mfano wa kujaza kwa kutumia mode.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

Kama tunavyoona, thamani ya null imebadilishwa. Bila shaka, tungeweza kuandika chochote badala ya `'True'` na kingechukua nafasi hiyo.


### Takwimu za Kihesabu
Sasa, tukija kwenye takwimu za kihesabu. Hapa, tuna njia mbili za kawaida za kujaza thamani zilizokosekana:

1. Kujaza kwa Median ya safu
2. Kujaza kwa Mean ya safu

Tunatumia Median pale ambapo data ina mwelekeo na ina outliers. Hii ni kwa sababu median haiguswi sana na outliers.

Wakati data imesawazishwa, tunaweza kutumia mean, kwa kuwa katika hali hiyo, mean na median zitakuwa karibu sana.

Kwanza, hebu tuchukue safu ambayo imegawanyika kawaida na tujaze thamani iliyokosekana kwa kutumia mean ya safu hiyo.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

Wastani wa safu ni


In [33]:
np.mean(fill_with_mean[0])

0.0

In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

Kama tunavyoona, thamani iliyokosekana imebadilishwa na wastani wake.


Sasa hebu tujaribu fremu nyingine ya data, na wakati huu tutabadilisha thamani za None na wastani wa safu.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

Median ya safu ya pili ni


In [36]:
fill_with_median[1].median()

4.0

In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

Kama tunavyoona, thamani ya NaN imebadilishwa na wastani wa safu hiyo


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

Unaweza kujaza nafasi zote tupu kwa thamani moja, kama `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> Mambo ya Muhimu:
1. Kujaza thamani zilizokosekana kunapaswa kufanywa pale ambapo kuna data kidogo au kuna mkakati wa kujaza data iliyokosekana.
2. Maarifa ya uwanja yanaweza kutumika kujaza thamani zilizokosekana kwa kuzikadiria.
3. Kwa data ya Kategoria, mara nyingi, thamani zilizokosekana hubadilishwa na hali ya kawaida ya safu husika.
4. Kwa data ya namba, thamani zilizokosekana kwa kawaida hujazwa na wastani (kwa seti za data zilizonormalishwa) au median ya safu husika.


In [40]:
# What happens if you try to fill null values with a string, like ''?


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

Unaweza pia **kujaza nyuma** ili kusambaza thamani halali inayofuata nyuma kujaza tupu:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

Kama unavyoweza kudhani, hii inafanya kazi vivyo hivyo na DataFrames, lakini pia unaweza kubainisha `axis` ambayo utaijaza thamani za null:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


Unaweza kuwa mbunifu kuhusu jinsi unavyotumia `fillna`. Kwa mfano, hebu tuangalie tena `example4`, lakini wakati huu tujaze thamani zilizokosekana kwa wastani wa thamani zote katika `DataFrame`:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

Kumbuka kwamba safu ya 3 bado haina thamani: mwelekeo wa chaguo-msingi ni kujaza thamani kwa mpangilio wa safu.

> **Mafunzo Muhimu:** Kuna njia nyingi za kushughulikia thamani zinazokosekana katika seti zako za data. Mkakati maalum unaotumia (kuondoa, kubadilisha, au hata jinsi unavyobadilisha) unapaswa kuamuliwa na maelezo ya data hiyo. Utapata uelewa bora wa jinsi ya kushughulikia thamani zinazokosekana kadri unavyoshughulikia na kuingiliana na seti za data.


### Usimbaji wa Data ya Kategoria

Miundo ya kujifunza kwa mashine hushughulikia tu nambari na aina yoyote ya data ya nambari. Haiwezi kutofautisha kati ya Ndiyo na Hapana, lakini inaweza kutofautisha kati ya 0 na 1. Kwa hivyo, baada ya kujaza thamani zilizokosekana, tunahitaji kusimba data ya kategoria kwa aina fulani ya nambari ili modeli iweze kuelewa.

Usimbaji unaweza kufanywa kwa njia mbili. Tutazijadili baadaye.


**KODIFIKESHENI YA LEBELE**  

Kodifikesheni ya lebele ni mchakato wa kubadilisha kila kategoria kuwa namba. Kwa mfano, tuseme tuna seti ya data ya abiria wa ndege na kuna safu inayojumuisha daraja lao miongoni mwa haya ['daraja la biashara', 'daraja la uchumi', 'daraja la kwanza']. Ikiwa kodifikesheni ya lebele itafanyika hapa, itabadilishwa kuwa [0,1,2]. Hebu tuone mfano kupitia msimbo. Kwa kuwa tutajifunza `scikit-learn` katika daftari zijazo, hatutaitumia hapa.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Ili kufanya usimbaji wa lebo kwenye safu ya kwanza, tunapaswa kwanza kuelezea ramani kutoka kwa kila darasa hadi nambari, kabla ya kubadilisha


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

Kama tunavyoona, matokeo yanalingana na tulivyotarajia. Kwa hivyo, ni lini tunatumia label encoding? Label encoding hutumika katika hali moja au zote zifuatazo:  
1. Wakati idadi ya kategoria ni kubwa  
2. Wakati kategoria ziko katika mpangilio.  


**UJUMUISHAJI WA ONE HOT**

Aina nyingine ya ujumishaji ni Ujumishaji wa One Hot. Katika aina hii ya ujumishaji, kila kategoria ya safu huongezwa kama safu tofauti, na kila kipengele cha data hupata 0 au 1 kulingana na kama kina kategoria hiyo au la. Kwa hivyo, ikiwa kuna kategoria n tofauti, safu n zitaongezwa kwenye dataframe.

Kwa mfano, hebu tuchukue mfano ule ule wa daraja la ndege. Kategoria zilikuwa: ['business class', 'economy class', 'first class']. Kwa hivyo, tukifanya ujumishaji wa one hot, safu tatu zifuatazo zitaongezwa kwenye seti ya data: ['class_business class', 'class_economy class', 'class_first class'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Tufanye usimbaji wa one hot kwenye safu ya kwanza


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

Kila safu iliyo na hot encoding ina 0 au 1, ambayo inaonyesha kama jamii hiyo ipo kwa data hiyo.


Tunatumia one hot encoding lini? One hot encoding hutumika katika mojawapo au zote za hali zifuatazo:

1. Wakati idadi ya kategoria na ukubwa wa seti ya data ni ndogo.
2. Wakati kategoria hazifuati mpangilio maalum.


> Mambo Muhimu:
1. Usimbaji hufanywa kubadilisha data isiyo ya nambari kuwa data ya nambari.
2. Kuna aina mbili za usimbaji: Usimbaji wa Lebo na Usimbaji wa One Hot, zote zinaweza kufanywa kulingana na mahitaji ya seti ya data.


## Kuondoa data inayojirudia

> **Lengo la kujifunza:** Kufikia mwisho wa sehemu hii, unapaswa kuwa na ujuzi wa kutambua na kuondoa thamani zinazojirudia kutoka kwa DataFrames.

Mbali na data inayokosekana, mara nyingi utakutana na data inayojirudia katika seti za data za maisha halisi. Kwa bahati nzuri, pandas inatoa njia rahisi ya kugundua na kuondoa maingizo yanayojirudia.


### Kutambua nakala: `duplicated`

Unaweza kutambua kwa urahisi thamani zinazojirudia kwa kutumia mbinu ya `duplicated` katika pandas, ambayo inarudisha maski ya Boolean inayoonyesha kama ingizo katika `DataFrame` ni nakala ya ingizo la awali. Hebu tuunde mfano mwingine wa `DataFrame` ili kuona hii ikifanya kazi.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### Kuondoa nakala: `drop_duplicates`
`drop_duplicates` inarudisha tu nakala ya data ambayo thamani zote za `duplicated` ni `False`:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

Zote `duplicated` na `drop_duplicates` kwa chaguo-msingi huzingatia safu zote lakini unaweza kubainisha kwamba zichunguze tu sehemu ndogo ya safu katika `DataFrame` yako:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2


---

**Kanusho**:  
Hati hii imetafsiriwa kwa kutumia huduma ya tafsiri ya AI [Co-op Translator](https://github.com/Azure/co-op-translator). Ingawa tunajitahidi kwa usahihi, tafadhali fahamu kuwa tafsiri za kiotomatiki zinaweza kuwa na makosa au kutokuwa sahihi. Hati ya asili katika lugha yake ya awali inapaswa kuzingatiwa kama chanzo cha mamlaka. Kwa taarifa muhimu, inashauriwa kutumia tafsiri ya kitaalamu ya binadamu. Hatutawajibika kwa maelewano mabaya au tafsiri zisizo sahihi zinazotokana na matumizi ya tafsiri hii.
